### ISSUE TITLE
A function for converting a Subset object to a Dataset object

### ISSUE THREAD

**Issue by Tomer-Barak** [pytorch/pytorch#97944]
### 🚀 The feature, motivation and pitch

After splitting it into train and test sets, I wanted to check how balanced a dataset is. However, random_split() returns Subset objects that don't have the attribute "targets." Instead, it requires manually using the indices attribute included in Subset. Today, converting a Subset to Dataset can be done manually using the indices or by constructing a Dataloader from the Subset and taking the Dataloader's dataset. Can you please create a simple function that converts a Subset to a Dataset?

### Alternatives

As an alternative, it is possible that random_split() would return two Datasets instead of two Subsets.

cc @VitalyFedyunin @ejguan @NivekT @dzhulgakov

**Comment by NivekT** [comment]
`Subset` currently inherits from `Dataset` class. Can you provide a code example?

**Comment by JerryLife** [comment]
Calling `dataset_train.dataset.targets` may not work because it would return all the targets instead of the targets of the subset. One may need to call 
```python
torch.tensor([dataset_train.dataset.targets[i] for i in dataset_train.indices])
```
to get the subset targets, which could be really slow. That is why we need a `Dataset` instead of a `Subset`.

**Comment by sanywright13** [comment]
sometimes you want to see the balance and distribution of your labels and targets in training set

**Comment by JulesCollenne** [comment]
Here's a simple workaround if you need to preserve access to these attributes (it gives them to the given Subset and returns it):

```
class SubsetWithSamples(Subset):
    def __init__(self, subset) -> None:
        super().__init__(subset.dataset, subset.indices)

        if hasattr(subset.dataset, "samples"):
            self.samples = [subset.dataset.samples[i] for i in subset.indices]
        else:
            self.samples = None

        if hasattr(subset.dataset, "targets"):
            self.targets = [subset.dataset.targets[i] for i in subset.indices]
        else:
            self.targets = None
```
Would love to see something like this natively supported in `torch.utils.data.Subset`.
